04-graphs/05-application.md

```diff
@@ -0,0 +1,704 @@
+            ])
+```
+
+##### !end-tests
+
+##### !hint
+Though the tests show the text representation of each maze, they will be converted into a graph before being passed to your function. You can assume that the start and end nodes will be valid locations in the maze.
+##### !end-hint
+
+##### !hint
+For the smaller mazes, you may wish to convert them to graphs by hand to help you visualize the graph representation. Try not to let the very large result path of the final maze intimidate you. Focus on the smaller mazes, which can be more easily visualized. The larger maze is included to help you test your code against a more complex maze, but as long as you can solve the smaller mazes, you should be able to solve the larger one as well.
+
+The next hint contains the steps from above, presented more closely to how they might be implemented in code. You may wish to try implementing the function without looking at the steps first, but if you get stuck, you can refer to the steps for guidance.
+##### !end-hint
+
+##### !hint
+```py
+def find_graph_path_helper(graph, start, end, visited, path):
+    # if we already visited the start node
+        # return None
+
+    # mark the start node as visited
+
+    # add the start node to the path
+
+    # if the start node is the end node
+        # return the path
+
+    # iterate over each of the neighbors of the start node
+        # recursively call our helper with the neighbor as the new start node
+        # if we got back a path
+            # return the path
+
+    # remove the start node from the end of the path
+
+    # return None
+```
+
+The next hint presents our implementation.
+##### !end-hint
+
+##### !hint
+```py
+def find_graph_path_helper(graph, start, end, visited, path):
+    # If the node we're about to visit has already been visited, we know this
+    # path doesn't lead to a solution
+    if start in visited:
+        return None
+    
+    # mark this node as now visited
+    visited.add(start)
+
+    # NEW - provisionally consider this node as part of the path
+    path.append(start)
+
+    # NEW - If the node is the end that we were looking for, we're done! Return
+    # the path that we constructed. This differs from basic depth first search
+    # which traverses through the entire graph. Here, we can stop as soon as we
+    # find the end node.
+    if start == end:
+        return path
+    
+    # If we reach this point, the node was not the end, but it may have
+    # neighbors for us to visit.
+    for move in graph[start]:
+        # Don't bother traversing to a node that's already been visited. This
+        # check isn't strictly necessary, but it can save a few recursive calls.
+        if move in visited:
+            continue
+
+        # Try to find a path from this adjacent node to the end. NEW - If a 
+        # path can be found, this call will return the path. If no path can be 
+        # found, it will return None.
```

Comment:
Would it be possible to leave a small comment indicating that the "move" parameter below is essentially the new "start" parameter. I could see some students getting confused by that part of this call.

Reply:
Sure. I can update the comments to say that the move node (though I think I'll rename it to next) becomes the start of the next attempted path search.